[Persistence] Consolidate common behaviour between `Pool` and `Account` into a shared interface (Issue #102)

h5law opened the pull request:
## Description

This PR addresses issue #102. It consolidates `Account` and `Pool` to use a shared interface `ProtocolAccountSchema` that generalises the functions that both actors use.

The `ProtocolAccountSchema` interface is implemented by the `baseProtocolAccountSchema` struct - following the same pattern laid out by the `ProtocolActorSchema` and `baseProtocolActorSchema`. Two new types are then created - `Account` and `Pool` these are instances of the `baseProtocolAccountSchema` struct and are initiated with their relevant `SQL` table names, height constraints and column names.

The functions in the `PersistenceRWContext` for both `Account` and `Pool` utilities are now generalised to use shared logic dramatically reducing the code footprint.

In addition to this all relevant unit tests have been updated and usages of these functions are working correctly. Also regarding issue #149 `address []bytes` has been removed from `InsertPool`.

## Issue

Fixes #102 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

- Create `ProtocolAccountSchema` interface and `baseProtocolAccountSchema` struct that implements this interface
- Create `Account` and `Pool` types that share the common functionality of `baseProtocolAccountSchema`
- Generalise all `Account` and `Pool` functions to use shared logic
- Remove `address []byte` from `InsertPool()`
- Seperate `shared_sql.go` and `types/shared_sql.go` into `account_shared_sql.go` and `actor_shared_sql.go` as well as `types/account_shared_sql.go` and `types/actor_shared_sql.go` respectively

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

h5law commented:
@Olshansk I have also addressed the discussion comment in the `persistence/account.go`  file and included the code for the relevant change if desired - commented out though. The current functionality is identical to how it was prior - only using shared logic instead.

Olshansk commented on persistence/shared_sql.go:
```diff
@@ -83,6 +86,19 @@ func (p *PostgresContext) GetActorsUpdated(actorSchema types.ProtocolActorSchema
 	return
 }
 
+func (p *PostgresContext) getAccountAmount(accountSchema types.ProtocolAccountSchema, address string, height int64) (amount string, err error) {
```
Can the functions here go into a separate `account_shared_sql`?

Might mean we have to rename this to `actor_shared_sql`

Olshansk commented on persistence/genesis.go:
```diff
@@ -46,8 +46,7 @@ func (m *persistenceModule) populateGenesisState(state *genesis.GenesisState) {
 	for _, pool := range state.GetPools() {
-		poolNameBytes := []byte(pool.GetAddress())
-		err = rwContext.InsertPool(pool.GetAddress(), poolNameBytes, pool.GetAmount())
+		err = rwContext.InsertPool(pool.GetAddress(), pool.GetAmount())
```
Please add a comment that the `address` for pools is actually the name.

You can find more details on it in [here](https://github.com/pokt-network/pocket-network-protocol/blob/issues/23/aat/utility/README.md#373-pool) or see the `genesis.json` file for reference:

```json
{
  "accounts": [
    {
      "address": "6f66574e1f50f0ef72dff748c3f11b9e0e89d32a",
      "amount": "100000000000000"
    },
  ],
  "pools": [
    {
      "address": "DAO",
      "amount": "100000000000000"
    },
  ],
```

h5law commented on persistence/shared_sql.go:
```diff
@@ -83,6 +86,19 @@ func (p *PostgresContext) GetActorsUpdated(actorSchema types.ProtocolActorSchema
+	if err != nil {
+		return
+	}
+	amount = defaultAccountAmountStr
```
@Olshansk Do you mean check the default is zero?
These two tests already exist: `TestDefaultNonExistentPoolAmount` and `TestDefaultNonExistentAccountAmount`

Or a standalone test for the `getAccountAmount` function? The other account/pool tests heavily use this function but I can add one if needed.

Olshansk replied:
Ah, I didn't know!

Olshansk commented on persistence/account.go:
```diff
@@ -28,14 +28,11 @@ func (p PostgresContext) SubtractAccountAmount(address []byte, amount string) er
-	//	return nil
-	// })
+	return p.operationAccountAmount(types.Account, hex.EncodeToString(address), amount, func(orig, amount *big.Int) error {
+		orig.Set(amount)
```
Any chance you can add a unit test for this? One for pool and one for account?

Olshansk commented on persistence/types/account_shared_sql.go:
```diff
@@ -0,0 +1,36 @@
+package types
+
```
```suggestion
// CLEANUP: Move SQL specific business logic into a `sql` package under `persistence`
```

@h5law Also, feel free to resolve comments (i.e. suggestions like these) after you apply them (either through github or yourself). There's always some inherent trust between the reviewer and reviewee.

Olshansk commented on persistence/types/base_account.go:
```diff
@@ -0,0 +1,52 @@
+// Implements the ProtocolAccountSchema interface that can be shared across both Accounts and Pools
+// allowing for the generalisation and sharing of code between these two entities
+type BaseProtocolAccountSchema struct {
+	// SQL Tables
```
Bump?

h5law replied:
Missed this one done now 👍🏼

Olshansk reviewed (approved):
@h5law LGTM! 

Just a couple things regarding review:

1. Make sure to re-request review when you need me to take another look
2. Don't hesitate to resolve the comments yourself if your confident you addressed the comment
3. Just double checking that LocalNet works fine? If so, go ahead and squash-and-merge!

h5law commented:
@Olshansk 

>     1. Make sure to re-request review when you need me to take another look
>     2. Don't hesitate to resolve the comments yourself if your confident you addressed the comment

I thought I did these but will make sure going forward 🫡
 
>     3. Just double checking that LocalNet works fine? If so, go ahead and squash-and-merge!

All good - will update CHANGELOGs and merge after that
